Add FXIOS-12758 [Summarizer] Add initial LiteLLMClient

issammani opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-12758)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/27778)

## :bulb: Description
This PR:
- Adds an initial `LiteLLMClient` to connect to openai like APIs.
- Adds both streaming and non-streaming modes.
- Adds support for different prompt roles ( to prevent prompt injection )

## :movie_camera: Demos
<!-- Please upload screenshots or video demos of your work, if applicable -->
<!-- You can either use a table (best for before/after screenshots) or the <details> disclosure -->

| Before | After |
| - | - |
| <!--insert "before" image here--> | <!--insert "after" image here--> |
| <!--insert "before" image here--> | <!--insert "after" image here--> |
| <!--insert "before" image here--> | <!--insert "after" image here--> |

<details>
<summary>Demo</summary>
<!-- Shorthand image template: <img height=400 src="<URL>" /> -->
</details>

## :pencil: Checklist
- [x] I filled in the ticket numbers and a description of my work
- [x] I updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] I ensured unit tests pass and wrote tests for new code
- [ ] If working on UI, I checked and implemented accessibility (Dynamic Text and VoiceOver)
- [ ] If adding telemetry, I read the [data stewardship requirements](https://github.com/mozilla-mobile/firefox-ios/wiki/Adding-Glean-Telemetry-Events) and will request a data review
- [ ] If needed, I updated documentation and added comments to complex code
- [ ] If needed, I added a backport comment (example `@Mergifyio backport release/v120`)

lmarceau commented on firefox-ios/Client.xcodeproj/project.pbxproj:
```diff
@@ -26893,6 +26898,7 @@
 				SWIFT_UPCOMING_FEATURE_IMPLICIT_OPEN_EXISTENTIALS = YES;
 				SWIFT_UPCOMING_FEATURE_IMPORT_OBJC_FORWARD_DECLS = YES;
 				SWIFT_UPCOMING_FEATURE_INFER_SENDABLE_FROM_CAPTURES = YES;
+				SWIFT_UPCOMING_FEATURE_REGION_BASED_ISOLATION = YES;
```
`SWIFT_UPCOMING_FEATURE_REGION_BASED_ISOLATION` shouldn't have changes?

issammani replied:
I hate xcode so so much 😢

lmarceau commented on firefox-ios/Client.xcodeproj/project.pbxproj:
```diff
@@ -27944,7 +27963,7 @@
 					"-Xlinker",
 					"-no_application_extension",
 				);
-				SWIFT_TREAT_WARNINGS_AS_ERRORS = YES;
+				SWIFT_TREAT_WARNINGS_AS_ERRORS = NO;
```
I am unsure why we see this change too here, since `SWIFT_TREAT_WARNINGS_AS_ERRORS` is `NO` on the main branch 🤔

issammani replied:
yeah my bad some commits crept in lol

lmarceau commented on firefox-ios/Client/Frontend/Settings/AppearanceSettings/Generic views/GenericSectionView.swift:
```diff
@@ -55,7 +55,7 @@ struct GenericSectionView<Content: View>: View {
 
             content()
-                .padding([.top, .bottom], UX.contentPadding)
+                // .padding([.top, .bottom], UX.contentPadding)
```
If it's commented it can be removed?

issammani replied:
No idea what happened. Lesson learned never use xcode thingy for commits again. Back to terminal lol

lmarceau reviewed (commented):
I was curious about the work and saw some stuff, so this is my spontaneous review 😄

issammani commented:
> I was curious about the work and saw some stuff, so this is my spontaneous review 😄
Thanks @lmarceau. Some commits crept in by mistake for some reason. I removed them. Should be good now. Also added your suggestions.

issammani commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/LiteLLMClientTests.swift:
```diff
@@ -0,0 +1,149 @@
+
+    override func setUp() {
+        super.setUp()
+        mockClient = LiteLLMClient(apiKey: Self.mockAPIKey, baseURL: URL(string: Self.mockAPIEndpoint)!)
```
Any examples we have anywhere I can copy from ?

issammani replied:
Done in https://github.com/mozilla-mobile/firefox-ios/pull/27779/commits/0f0999d945dce7517e5d0add0a678a1598801a88

cyndichin commented on firefox-ios/Client/Frontend/Summary/LiteLLMClient.swift:
```diff
@@ -0,0 +1,235 @@
+        stream: Bool = false,
+        completion: ((Result<String, Error>) -> Void)? = nil
+    ) {
+        do {
+            let request = try makeRequest(messages: messages, model: model, maxTokens: maxTokens, stream: stream)
+            if stream {
+                session.dataTask(with: request).resume()
+            } else {
+                URLSession.shared.dataTask(with: request) { data, _, error in
+                    self.handleResponse(data: data, error: error, completion: completion)
+                }.resume()
+            }
+        } catch {
+            if stream {
+                streamDelegate?.liteLLMClient(self, didFinishWith: LLMClientError.requestCreationFailed)
+            } else {
+                completion?(.failure(LLMClientError.requestCreationFailed))
+            }
+        }
+    }
```
+1 since we're moving towards adopting modern concurrency, we can look into using the async / await version of url session - https://www.avanderlee.com/concurrency/urlsession-async-await-network-requests-in-swift/

issammani replied:
I plan to do this in a follow-up. We just had issues recently dealing with `Task{}` stuff, I don't want that to happen here for the MVP. Will open a follow-up ticket to make everything async.

cyndichin commented on firefox-ios/Client/Frontend/Summary/LiteLLMClient.swift:
```diff
@@ -0,0 +1,162 @@
+    /// - Parameters:
+    ///   - messages: Array of `ChatMessage`.
+    ///   - model: Model identifier.
+    ///   - maxTokens: Maximum tokens to generate.
```
would love an explanation of the `1000` as well

issammani replied:
Definitely will add a comment here. But this has to do with budget limits.

issammani replied:
Updated [here](https://github.com/mozilla-mobile/firefox-ios/pull/27779/commits/b49f45a9727f06a605cdfd65ceab62325dd6626a) to use a separate struct and no defaults for now. I will add a comment at the call site when the struct is used about the 1000 limit. I think this is cleaner instead of having magic values scattered all over as I did before 😄 Let me know what you think !

cyndichin replied:
perfect! like this so much better, nice job!!

cyndichin commented on firefox-ios/Client/Frontend/Summary/LiteLLMClientError.swift:
```diff
@@ -0,0 +1,29 @@
+    case decodingFailed
+    case networkError(underlying: Error)
+
+    public var errorDescription: String? {
```
```suggestion
    public var errorDescription: String {
```
will this ever be nil if we cover all cases?

issammani replied:
Updated in https://github.com/mozilla-mobile/firefox-ios/pull/27779/commits/09bfb49e63018a1e72b8cfe4128433b13fc63031

cyndichin commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/LiteLLMClientTests.swift:
```diff
@@ -0,0 +1,148 @@
+import Common
+import Shared
+
+class LiteLLMClientTests: XCTestCase {
```
```suggestion
final class LiteLLMClientTests: XCTestCase {
```

thanks for writing tests!! It also makes it easier to see what the payloads / responses will look like. Would be nice to test all the other `public` methods as well if we can.

cyndichin reviewed (commented):
thanks @issammani! added my thoughts for the initial review, let me know what you think or if anything isn't clear!

cyndichin commented on firefox-ios/Client/Frontend/Summary/LiteLLMRole.swift:
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+/// Defines the role of an LLM message, catching typos at compile time.
+public enum LiteLLMRole: String, Codable {
```
Should we add about prompt injection here? 🤔

issammani commented on firefox-ios/Client/Frontend/Summary/LiteLLMClient.swift:
```diff
@@ -0,0 +1,162 @@
+    private var logger: Logger = DefaultLogger.shared
+    private let apiKey: String
+    private let baseURL: URL
+    private lazy var session: URLSession = {
```
Added follow-up comment and ticket about `URLSessionProtocol`  in https://github.com/mozilla-mobile/firefox-ios/pull/27779/commits/f506279a446248ce65defe0cdb1798b29997ffa3
